unirec/model/rank/adaranker.py

```diff
@@ -0,0 +1,196 @@
+import torch
+import torch.nn as nn
+
+from ..base.ranker import Ranker
+import unirec.model.modules as modules
+
+
+class AdaRanker(Ranker):
+    def __init__(self, config):
+        self.train_type = config['train_type']
+        self.base_model = config['base_model']
+        assert self.base_model in ["GRU", "SASRec"], "we only support 'GRU' and 'SASRec' as base model in AdaRanker now."
+
+        if self.base_model == "SASRec":
+            self.n_layers = config['n_layers']
+            self.n_heads = config['n_heads']
+            self.inner_size = config['inner_size']  # the dimensionality in feed-forward layer
+            self.hidden_dropout_prob = config['hidden_dropout_prob']
+            self.attn_dropout_prob = config['attn_dropout_prob']
+            self.hidden_act = config['hidden_act']
+            self.layer_norm_eps = config['layer_norm_eps']
+            self.max_seq_len = config['max_seq_len']
+            self.use_pos_emb = config['use_position_emb']
+
+        super(AdaRanker, self).__init__(config)
+
+        self.dnn_input_size = self.embedding_size * 2
+        self.dnn_inner_size = self.embedding_size
+
+        if self.train_type == 'Ada-Ranker':
+            # extract distribution layers
+            self.extract_distribution_layer = modules.NeuProcessEncoder(
+                self.embedding_size, self.embedding_size, self.embedding_size, 
+                self.dropout_prob, self.device
+            )
+
+            # add bias layers
+            self.film_affine_emb_scale = nn.Linear(self.embedding_size, 1)
+            self.film_affine_emb_bias = nn.Linear(self.embedding_size, 1)
+
+            # predict_layer
+            self.mem_w1 = modules.MemoryUnit(self.dnn_input_size, self.dnn_inner_size, self.embedding_size)
```

Comment:
why hardcode 2 memory units?

Reply:
cause self.mlp_layers is a two-layer MLP, we need 2 memory units for each layer as w_add and b_add.